README.md
# ImageEngine Angular [![Build Status](https://travis-ci.org/WURFL/ImageEngine-angular.png)](https://travis-ci.org/WURFL/ImageEngine-angular)

> An AngularJS directive for ImageEngine

# ImageEngine

[ImageEngine](http://www.scientiamobile.com/page/imageengine?utm_source=npmjs.com&utm_medium=page&utm_term=angular-component&utm_campaign=angular-component) is an intelligent image CDN for optimizing, compressing and resizing images. ImageEngine will enhance your responsive images by enabling support for HTTP2, automatic webp conversion, Client Hints and more.

It is likely that ImageEngine will shave off 50-60% of data traffic generated by images on your site. To better support mobile devices, ImageEngine relies on WURFL for server side device detection in cases where responsive images and Client Hints fall short.

[Registration](https://scientiamobile.com/imageengine/signup?utm_source=npmjs.com&utm_medium=page&utm_term=angular-component&utm_campaign=angular-component#imageengine-lite) is required to get the most out of ImageEngine. 

Images weight is by far the most important challenge to address when optimizing a web page for the plethora of devices on the web today. This is why ImageEngine will instantly give your users a better experience by reducing load time of your page. Not to mention the reduced data cost.

ImageEngine works as a CDN proxy. By referencing the ImageEngine servers and providing the full URL to the original image as a parameter, ImageEngine will identify the device, and its capabilities, and resize and optimize the image accordingly.

## Client Hints

The plugin will enable Client Hint support for your site. Client Hints, with information about the viewport size, device pixel ratio and actual image size, are added to the image request (HTTP header) enabling ImageEngine to resize the image with surgical precision.

## HTTP2 support

HTTP will give additional performance improvement on the network level. ImageEngine supports HTTP2 over SSL, which means that if you serve your sites and images by `https://`, ImageEngine will automatically serve all images over HTTP2.

## WebP

WebP is a lightweight image format with great quality. WebP is well supported by browsers. ImageEngine will detect if the end user's browser supports WebP and automatically convert any format to WebP to increase performance.

## Usage

### Install with yarn
* `yarn add image-engine-angular`
* Include `image-engine-angular.js`. It should be located at `node_modules/src/image-engine-angular.js`

### Install with npm
* `npm i image-engine-angular --save`
* Include `image-engine-angular.js`. It should be located at `node_modules/src/image-engine-angular.js`

### Install with bower
* `bower install image-engine-angular --save`
* Include `image-engine-angular.js`. It should be located at `bower_components/src/image-engine-angular.js`

### Install from source
* [Download Latest Version](https://github.com/WURFL/ImageEngine-angular/releases) and extract the archive.
* Include `image-engine-angular.js`. It should be located at `archive_path/src/image-engine-angular.js`

### How to use it

* Include the image-engine-angular directive dependency on your angular module:

```
    var app = angular.module("demoapp", ["image-engine-angular"]);
    app.config(function (imgEngConfigProvider) {
       imgEngConfigProvider.setToken('your-token');
    });
```

Note: [sign up here](https://scientiamobile.com/imageengine/signup?utm_source=npmjs.com&utm_medium=page&utm_term=angular-component&utm_campaign=angular-component#imageengine-lite) to get your token

* Include the markup directive on your HTML page, like this:

`<img-eng src="http://yourserver.com/image.png"></img-eng>`

or if you want to use interpolation:

`<img-eng src="{{myUrl}}"></img-eng>`

where {{myUrl}} is the url of the [trusted image](https://docs.angularjs.org/api/ng/service/$sce) to load.

## Examples

Check the [ImageEngine Documentation](https://docs.scientiamobile.com/documentation/image-engine/image-engine-getting-started) for the list of available settings.

### Fully Automatic

`<img-eng src="http://yourserver.com/image.png"></img-eng>`

### 20% of screen size

`<img-eng src="http://yourserver.com/image.png" pc="20"></img-eng>`

### Create an image 300px Wide

`<img-eng src="http://yourserver.com/image.png" w="300"></img-eng>`

### Create 200x200px Thumbnails with Black Letterboxes/Pillarboxes

`<img-eng src="http://yourserver.com/image.png" w="200" h="200" m="letterbox_000000_100"></img-eng>`

## Demo

* Run: `npm start`
* Browse: `http://localhost:8000/demo/index.html`

## Testing

``` bash
$ npm test
```

## Authors

- [Luca Corbo](https://github.com/lucor) — [view contributions](https://github.com//WURFL/image-engine-angular/commits?author=lucor)
- [All Contributors](../../contributors)

## License

Licensed under the MIT license. (See the LICENSE file)

Copyright &copy; ScientiaMobile, Inc.


CHANGELOG.md
# Changelog

All notable changes to `image-engine-angular` will be documented in this file

## 1.0.3 - 2017-08-01

### Updated
- Defines AngularJS as peer dependency

## 1.0.2 - 2017-07-20

### Added
- Crop (`cr`) parameter support
- Compression (`cmpr`) parameter support
- Inline (`in`) parameter support
- Tests for AngularJS versions 1.6.x

### Deprecated
- isLite() method

## 1.0.1 - 2016-03-09

### Added
- Tests for AngularJS versions 1.5.x

### Updated
- Dev dependencies

### Fixed
- strict mode error: function(imgEngConfig) is not using explicit annotation and cannot be invoked in strict mode (#4)

## 1.0.0 - 2015-11-16

### Changed
- WURFL Image Tailor (WIT) is now ImageEngine

### Added
- Token support
- Flag to switch between "ImageEngine" and "ImageEngine Lite"

### Updated
- Base Urls
- Tests

## 0.9.4 - 2015-03-23

### Added
- $observe expression on all allowed attributes

## 0.9.3 - 2015-02-27

### Fixed
- When ngSrc was undefined at the beginning, image wasn't refreshed

## 0.9.2 - 2015-02-26

### Added
- $observe expression on 'ngSrc'
- CHANGELOG file
- Gulp.js support
- Tests for AngularJS versions 1.2.x, 1.3.x, 1.4.x

## 0.9.1 - 2015-01-19

### Added
- ngSrc directive support

### Changed
- copyright info

## 0.9.0 - 2014-06-20

### Added
- Initial release


test/unit/providerSpec.js
describe("Unit: Testing angular provider for ImageEngine", function() {

    var provider;

    beforeEach(module('image-engine-angular', function(imgEngConfigProvider) {
        provider = imgEngConfigProvider;
    }));

    it('tests the providers internal function', inject(function() {
        provider.setToken('a-token');
        expect(provider.$get().token).to.be.equal('a-token');
        expect(provider.$get().is_lite).to.be.false;
        provider.isLite();
        expect(provider.$get().is_lite).to.be.true;
    }));

});


test/unit/directiveSpec.js
describe("Unit: Testing angular directive for ImageEngine", function() {

    var scope, compile, provider;

    beforeEach(module('image-engine-angular', function(imgEngConfigProvider) {
        provider = imgEngConfigProvider;
        provider.setToken('token');
    }));

    beforeEach(inject(
        ['$compile','$rootScope', '$sce', function($c, $r, $sce) {
            compile = $c;
            scope = $r;
            sce = $sce;
        }]
    ));

    it('should work as an element', function () {
        var elt = angular.element('<img-eng></img-eng>');
        compile(elt)(scope);
        scope.$digest();

        expect(elt.html()).to.be.a('string');
        expect(elt.html()).to.be.equal('<div class="img-eng"><img ng-src=""></div>');
    });

    it('should work as an element with image source', function () {
        var elt = angular.element('<img-eng src="http://test.com/image.jpg"></img-eng>');
        compile(elt)(scope);
        scope.$digest();

        expect(elt.html()).to.be.a('string');
        expect(elt.html()).to.be.equal('<div class="img-eng"><img ng-src="//token.imgeng.in/http://test.com/image.jpg" src="//token.imgeng.in/http://test.com/image.jpg"></div>');
    });

    it('should work as an element with image source and param attribute', function () {
        var elt = angular.element('<img-eng src="http://test.com/image.jpg" w="200"></img-eng>');
        compile(elt)(scope);
        scope.$digest();

        expect(elt.html()).to.be.a('string');
        expect(elt.html()).to.be.equal('<div class="img-eng"><img ng-src="//token.imgeng.in/w_200/http://test.com/image.jpg" src="//token.imgeng.in/w_200/http://test.com/image.jpg"></div>');
    });

    it('should work as an element with ngSrc directive', function () {
        var elt = angular.element('<img-eng ng-src="http://test.com/image.jpg"></img-eng>');
        compile(elt)(scope);
        scope.$digest();

        expect(elt.html()).to.be.a('string');
        expect(elt.html()).to.be.equal('<div class="img-eng"><img ng-src="//token.imgeng.in/http://test.com/image.jpg" src="//token.imgeng.in/http://test.com/image.jpg"></div>');
    });

    it('should work as an element with ngSrc directive and a trusted resource', function () {
        scope.myUrl = sce.trustAsResourceUrl('http://test.com/image.jpg');
        var elt = angular.element('<img-eng ng-src="{{myUrl}}" w="200"></img-eng>');
        compile(elt)(scope);
        scope.$digest();

        expect(elt.html()).to.be.a('string');
        expect(elt.html()).to.be.equal('<div class="img-eng"><img ng-src="//token.imgeng.in/w_200/http://test.com/image.jpg" src="//token.imgeng.in/w_200/http://test.com/image.jpg"></div>');
    });

    it('should change when ngSrc directive variable change', function () {
        scope.myUrl = sce.trustAsResourceUrl('http://test.com/image.jpg');
        var elt = angular.element('<img-eng ng-src="{{myUrl}}"></img-eng>');
        compile(elt)(scope);
        scope.$digest();
        scope.myUrl = sce.trustAsResourceUrl('http://test.com/image2.jpg');
        scope.$digest();

        expect(elt.html()).to.be.a('string');
        expect(elt.html()).to.be.equal('<div class="img-eng"><img ng-src="//token.imgeng.in/http://test.com/image2.jpg" src="//token.imgeng.in/http://test.com/image2.jpg"></div>');
    });

    it('should change when ngSrc directive variable change from undefined', function () {
        scope.myUrl = undefined;
        var elt = angular.element('<img-eng ng-src="{{myUrl}}"></img-eng>');
        compile(elt)(scope);
        scope.$digest();
        scope.myUrl = sce.trustAsResourceUrl('http://test.com/image2.jpg');
        scope.$digest();

        expect(elt.html()).to.be.a('string');
        expect(elt.html()).to.be.equal('<div class="img-eng"><img ng-src="//token.imgeng.in/http://test.com/image2.jpg" src="//token.imgeng.in/http://test.com/image2.jpg"></div>');
    });

    it('should ignore invalid attribute', function () {
        var elt = angular.element('<img-eng src="http://test.com/image.jpg" w="200" in="true" a="invalid"></img-eng>');
        compile(elt)(scope);
        scope.$digest();

        expect(elt.html()).to.be.a('string');
        expect(elt.html()).to.be.equal('<div class="img-eng"><img ng-src="//token.imgeng.in/w_200/in_true/http://test.com/image.jpg" src="//token.imgeng.in/w_200/in_true/http://test.com/image.jpg"></div>');
    });

});


test/angular-1.2/gulpfile.js
var gulp = require('gulp');
var karma = require('karma').server;
var bower = require('gulp-bower');

/**
 * Install dependencies for angular-1.2.x
 */
gulp.task('bower-1.2', function() {
    return bower({ directory: './bower_components', cwd: './test/angular-1.2' })
});

/**
 * Run tests for angular-1.2.x
 */
gulp.task('test-1.2', ['bower-1.2'], function (done) {
    karma.start({
        configFile: __dirname + '/karma.conf.js',
        singleRun: true,
        noAutoWatch: true,
        reporters: 'dots',
        browsers: ['PhantomJS']
    }, done);
});

karma.conf.js
// Karma configuration

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['mocha','chai'],


    // list of files / patterns to load in the browser
    files: [

      'bower_components/angular/angular.js',
      'bower_components/angular-mocks/angular-mocks.js',
      'src/image-engine-angular.js',
      'test/**/*Spec.js'
    ],


    // list of files to exclude
    exclude: [
      
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
    
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['PhantomJS'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false
  });
};


src/image-engine-angular.js
'use strict';
/*
 * image-engine-angular
 * An AngularJS directive for ImageEngine
 * Homepage: https://github.com/WURFL/ImageEngine-angular
 * Authors: Luca Corbo (https://github.com/lucor)
 * (c) 2014 - 2017 ScientiaMobile, Inc.
 * License: MIT
 */

angular.module('image-engine-angular', [])
    .provider('imgEngConfig', function () {
        this.is_lite = false;
        this.token = null;
        this.setToken = function (token) {
            this.token = token;
        };
        this.isLite = function () {
            // Deprecated method since 1.0.2
            this.is_lite = true;
        };
        this.$get = function () {
            return this;
        };
    })
    .factory('imgEngUrls', ['imgEngConfig', function (imgEngConfig) {
        var base_url = ['//' + imgEngConfig.token];

        if (imgEngConfig.is_lite === true) {
            base_url.push('lite');
        }

        base_url.push('imgeng.in');

        return {
            get: function () {
                return base_url.join('.');
            }
        };
    }])
    .directive('imgEng', ['imgEngUrls', function (imgEngUrls) {
        return {
            restrict: 'E',
            replace: false,
            scope: {
                src: '@',
                w: '@',
                h: '@',
                pc: '@',
                m: '@',
                f: '@',
                r: '@',
                cr: '@',
                cmpr: '@',
                in: '@'
            },
            template: function (element, attributes) {
                return '<div class="img-eng"><img ng-src="{{imgeng_link}}"/></div>';
            },
            link: function (scope, element, attributes) {
                scope.imgeng_link = '';

                var allowedAttributes = ['w', 'h', 'pc', 'm', 'f', 'r', 'cr', 'cmpr', 'in'];

                scope.$watchCollection('[src, w, h, pc, m, f, r, cr, cmpr, in]', function (values, oldValues) {

                    var src = values.shift();
                    if (!src) {
                        return;
                    }

                    var imgeng_link_pieces = [imgEngUrls.get()];
                    angular.forEach(allowedAttributes, function (attribute, index) {
                        if (values[index]) {
                            imgeng_link_pieces.push(attribute + '_' + values[index]);
                        }
                    });

                    imgeng_link_pieces.push(src);
                    scope.imgeng_link = imgeng_link_pieces.join('/');
                });
            }
        };
    }]);


src/angular-wurfl-image-tailor.js
/*
 * angular-wurfl-image-tailor
 * Authors: Luca Corbo (https://github.com/lucor)
 * (c) 2014 - 2015 ScientiaMobile, Inc.
 * License: MIT
 */

angular.module('angular-wurfl-image-tailor', [])
    .factory('witUrls', function () {
        var wit_index = 0;
        var wit_urls = ['//wit.wurfl.io', '//wit1.wurfl.io', '//wit2.wurfl.io', '//wit3.wurfl.io', '//wit4.wurfl.io'];
        return {
            get: function () {
                var url = wit_urls[wit_index];
                wit_index++;
                if (wit_index >= wit_urls.length) {
                    wit_index = 0;
                }
                return url;
            }
        };
    })
    .directive('imgWit', ['witUrls', function (witUrls) {
        return {
            restrict: 'E',
            replace: false,
            scope: {
                src: '@',
                w: '@',
                h: '@',
                pc: '@',
                m: '@',
                f: '@',
                r: '@'
            },
            template: function (element, attributes) {
                return '<div class="wit"><img ng-src="{{wit_link}}"/></div>';
            },
            link: function (scope, element, attributes) {
                scope.wit_link = '';

                var allowedAttributes = ['w', 'h', 'pc', 'm', 'f', 'r'];

                scope.$watchCollection('[src, w, h, pc, m, f, r]', function (values, oldValues) {

                    var src = values.shift();
                    if (!src) {
                        return;
                    }

                    var wit_link_pieces = [witUrls.get()];
                    angular.forEach(allowedAttributes, function (attribute, index) {
                        if (values[index]) {
                            wit_link_pieces.push(attribute + '_' + values[index]);
                        }
                    });

                    wit_link_pieces.push(src);
                    scope.wit_link = wit_link_pieces.join('/');
                });
            }
        }
    }]);


gulpfile.js
var gulp = require('gulp');
var karma = require('karma').Server;

require('./test/angular-1.2/gulpfile');
require('./test/angular-1.3/gulpfile');
require('./test/angular-1.4/gulpfile');
require('./test/angular-1.5/gulpfile');
require('./test/angular-1.6/gulpfile');

/**
 * Run tests for all angular versions
 */
gulp.task('test-all-versions', ['test-1.2', 'test-1.3', 'test-1.4', 'test-1.5', 'test-1.6'], function (done) {
});

/**
 * Run test once and exit
 */
gulp.task('test', function (done) {
    karma.start({
        configFile: __dirname + '/karma.conf.js',
        singleRun: true
    }, done);
});

/**
 * Watch for file changes and re-run tests on each change
 */
gulp.task('tdd', function (done) {
    karma.start({
        configFile: __dirname + '/karma.conf.js'
    }, done);
});

gulp.task('default', ['tdd']);
